metadata-ingestion/src/datahub/ingestion/source/nifi.py

```diff
@@ -196,6 +197,73 @@ def validator_site_url(cls, site_url: str) -> str:
         return site_url
 
 
+class BidirectionalComponentGraph:
+    def __init__(self):
+        self.outgoing: Dict[str, Set[str]] = defaultdict(set)
+        self.incoming: Dict[str, Set[str]] = defaultdict(set)
+        # this will not count duplicates/removal of non-existing connections correctly - it is only for quick check
+        self.connections_cnt = 0
+
+    def add_connection(self, from_component: str, to_component: str) -> None:
+        # this is sanity check
+        outgoing_duplicated = to_component in self.outgoing[from_component]
+        incoming_duplicated = from_component in self.incoming[to_component]
+
+        self.outgoing[from_component].add(to_component)
+        self.incoming[to_component].add(from_component)
+        self.connections_cnt += 1
+
+        if outgoing_duplicated or incoming_duplicated:
+            logger.warning(
+                f"Somehow we attempted to add a connection between 2 components which already existed! Duplicated incoming: {incoming_duplicated}, duplicated outgoing: {outgoing_duplicated}. Connection from component: {from_component} to component: {to_component}"
+            )
+
+    def remove_connection(self, from_component: str, to_component: str) -> None:
+        self.outgoing[from_component].discard(to_component)
+        self.incoming[to_component].discard(from_component)
+        self.connections_cnt -= 1
+
+    def get_outgoing_connections(self, component: str) -> Set[str]:
+        return self.outgoing[component]
+
+    def get_incoming_connections(self, component: str) -> Set[str]:
+        return self.incoming[component]
+
+    def delete_component(self, component: str) -> None:
+        logger.debug(f"Deleting component with id: {component}")
+        incoming = self.incoming[component]
+        logger.debug(
+            f"Recognized {len(incoming)} incoming connections to the component"
+        )
+        outgoing = self.outgoing[component]
+        logger.debug(
+            f"Recognized {len(outgoing)} outgoing connections from the component"
+        )
+
+        for i in incoming:
+            for o in outgoing:
+                self.add_connection(i, o)
+
+        for i in incoming:
+            self.outgoing[i].remove(component)
+        for o in outgoing:
+            self.incoming[o].remove(component)
+
+        added_connections_cnt = len(incoming) * len(outgoing)
+        deleted_connections_cnt = len(incoming) + len(outgoing)
+        logger.debug(
+            f"Deleted {deleted_connections_cnt} connections and added {added_connections_cnt}"
+        )
+
+        del self.outgoing[component]
+        del self.incoming[component]
+
+        self.connections_cnt -= deleted_connections_cnt
```

Comment:
```suggestion
        self.connections_cnt +=  added_connections_cnt - deleted_connections_cnt
```

Reply:
This would be wrong - the `add_connection` function already increments the counter.

Reply:
Oh, can you add a comment mentioning so , or use `remove_connection`?

Reply:
The count thing wasn't so clear to me.

Reply:
I added an explanatory comment, `remove_connection` shouldn't be used while removing a `component` (due to performance reasons)